---
title: "`r params$exposure` posthoc analysis"
date: '`r paste("Updated on", Sys.Date())`'
output: 
  html_document:
    css: posthoc/posthoc_report.css
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
    toc_depth: 3
    theme: "paper"
params:
  exposure: "diab"
  covariates: !r c("age_ref_imp", "sex", "pc1", "pc2", "pc3", "study_gxe")
  hrc_version: "v2.3"
  path: "/media/work/gwis_test"
---

  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, message = F, warning = F, error = F)

# load packages and define wrapper functions
library(tidyverse)
library(data.table)
library(forcats)
library(ggplot2)
library(qqman)
library(rlang)
library(figifs)
library(DT)
library(grid)
library(stargazer)
library(kableExtra)
library(formattable)
library(glue)
```

# Description

## GWIS results followup
- Table of significant and suggestive (1df GxE only) ```r params$exposure``` GWIS findings  
<!-- - Plots of predicted logodds stratified by exposure and genotype () -->
<!--   - *outcome ~ snp_dose:exposure + covariates*   -->
<!--   - assumes mean values for adjustment covariates    -->
- GxE models stratified by study_design and tumor site (proximal, distal, rectal)  
- GxE models with additional covariate sets (if requested)  
- Odds ratios stratified by exposure and genotype  
  <!-- - model of the form *outcome ~ exposure + covariates + snp_p1 + snp_p2 + exposure:covariates + exposure:snp_p1 + exposure:snp_p2* -->
  - Odds ratios with single reference group (E = 0, G = 0), and exposure odds ratios stratified by imputed genotype  
    - model: *outcome ~ exposure + covariates + snp_p1 + snp_p2 + exposure:snp_p1 + exposure:snp_p2*  
  - per risk allele odds ratios stratified by exposure
    - model:  *outcome ~ exposure + covariates + dosage + exposure:dosage*  
  - **exposures and SNPs coded to reflect risk increase**  
- Locuszoom plot  
  - based on statistic that identified the hit (e.g. gxe, 2df, 3df, etc)
- Functional annotation plot  

## Covariates Sets
- **Set1**: age_ref_imp, sex, energytot_imp, study_gxe, pc1, pc2, pc3

------

<!-- # SNP information -->
<!-- ```{r} -->
<!-- child_env <- new.env() -->
<!-- assign("exposure", params$exposure, child_env) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- include_methods <- c("manhattan_chiSqGxE_aspirin_clump_df",  -->
<!--                      "twostep_wht_chiSqG_aspirin_expectation_hybrid_df") -->

<!-- x <- readRDS(glue("/media/work/gwis/posthoc/{child_env$exposure}/gwis_sig_results_input_{child_env$exposure}.rds")) %>% -->
<!--   dplyr::mutate(SNP = paste0("chr", gsub("\\:", "_", SNP)), -->
<!--                 AAF = round(GxESet_AltAlleleFreq, 3), -->
<!--                 Rsq = round(GxESet_Rsq, 3)) %>% -->
<!--   dplyr::filter(method %in% include_methods) %>%  -->
<!--   dplyr::rename(Chr = Chromosome, -->
<!--                 BP = Location) %>% -->
<!--   arrange(method, Chr, BP) %>% -->
<!--   dplyr::select(SNP, Chr, BP, method, AAF, Rsq) -->

<!-- kable(x, caption = "Significant GxE SNPs", escape = F) %>% -->
<!--   kable_styling("striped", full_width = F, position = 'center') -->
<!-- ``` -->

<!-- define variables -->
```{r, include = F}
child_env <- new.env()
assign("exposure",        params$exposure, child_env)
assign("hrc_version",     params$hrc_version, child_env)
assign("covariates", sort(params$covariates), child_env)
# assign("method", "chiSqGxE", child_env)

# exposure working directory path
path = glue("/media/work/gwis_test/{params$exposure}")
```




# GxE (1df)
```{r, results = 'asis'}
assign("method", "chiSqGxE", child_env)

# snps_out <- readRDS(glue("{output_dir}manhattan_chiSqGxE_{params$exposure}_clump_df.rds")) %>%
#   dplyr::filter(chiSqGxE_p <= 5e-8) %>%
#   dplyr::arrange(Chromosome, Location) %>%
#   dplyr::mutate(snps = paste0('chr', gsub("\\:", "\\_", SNP))) %>%
#   dplyr::pull(snps)

snps_out <- paste0('chr', gsub(":", "_", c("6:12577203:T:C", "6:32560631:C:T")))

res <- lapply(snps_out, function(snp) {
  assign("snp", snp, child_env)
  knitr::knit_child('posthoc/posthoc_child.Rmd', envir = child_env, quiet = TRUE)
  }
)

cat(unlist(res), sep = '\n')
```


<!-- # Suggestive GxE findings -->
<!-- ```{r} -->
<!-- child_env <- new.env() -->
<!-- assign("exposure", params$exposure, child_env) -->
<!-- assign("method", "chiSqGxE", child_env) -->
<!-- assign("covariates", sort(covariates_set1), child_env) -->
<!-- ``` -->

<!-- ```{r, results = 'asis'} -->
<!-- snps_out <- readRDS(glue("{output_dir}manhattan_chiSqGxE_{params$exposure}_clump_df.rds")) %>% -->
<!--   dplyr::arrange(Chromosome, Location) %>% -->
<!--   dplyr::mutate(snps = paste0('chr', gsub("\\:", "\\_", SNP))) %>% -->
<!--   dplyr::pull(snps) -->


<!-- res <- lapply(snps_out, function(snp) { -->
<!--   assign("snp", snp, child_env) -->
<!--   knitr::knit_child('posthoc_child.Rmd', envir = child_env, quiet = TRUE) -->
<!--   } -->
<!-- ) -->

<!-- cat(unlist(res), sep = '\n') -->
<!-- ``` -->





# Two-step finding (expectation based)
```{r, results = 'asis'}

assign("method", "chiSqGxE", child_env)

# snps_out <- readRDS(glue(output_dir, "twostep_wht_chiSqEDGE_{params$exposure}_expectation_hybrid_df.rds")) %>%
#   dplyr::filter(SNP == '5:40252294:C:T' | SNP == "6:32560631:C:T") %>%
#   arrange(Chromosome, Location) %>%
#   dplyr::mutate(snps = paste0('chr', gsub("\\:", "\\_", SNP))) %>%
#   pull(snps)

snps_out <- paste0('chr', gsub(":", "_", c("5:40252294:C:T")))

res <- lapply(snps_out, function(snp) {
  assign("snp", snp, child_env)
  knitr::knit_child(
    glue('posthoc/posthoc_child.Rmd'), envir = child_env, quiet = TRUE)
  }
)
cat(unlist(res), sep = '\n')
```



