---
title: "`r params$exposure` posthoc analysis"
date: '`r paste("Updated on", Sys.Date())`'
output: 
  html_document:
    css: posthoc/posthoc_report.css
    toc: true
    toc_float:
      collapsed: true
      smooth_scroll: false
    toc_depth: 3
    theme: "paper"
params:
  exposure: "bmi5"
  covariates: !r c("age_ref_imp", "sex", "pc1", "pc2", "pc3", "study_gxe")
  hrc_version: "v2.3"
  path: "/media/work/gwis_test"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, message = F, warning = F, error = F)

# load packages and define wrapper functions
library(tidyverse)
library(data.table)
library(forcats)
library(ggplot2)
library(qqman)
library(rlang)
library(figifs)
library(DT)
library(grid)
library(stargazer)
library(kableExtra)
library(formattable)
library(glue)
library(qs)
```

# Description

## Follow-up analyses of GWIS findings  
- Table of significant and suggestive (1df GxE only) ```r params$exposure``` GWIS findings  
- GxE models stratified by study_design and tumor site (proximal, distal, rectal)  
- GxE models with additional covariate sets (might do this for aspirin ..)  
- Odds ratios stratified by exposure and genotype  
- Plots of model predicted outcome by 1 unit increase of exposure/allelic dosage  
- Locuszoom plot   
- Functional annotation plot   

---

<!-- define variables -->
```{r, include = F}
child_env <- new.env()
assign("exposure",        params$exposure, child_env)
assign("hrc_version",     params$hrc_version, child_env)
assign("covariates", sort(params$covariates), child_env)

# exposure working directory path
path = glue("/media/work/gwis_test/{params$exposure}/")
```




# GxE 1df (p < 5e-6)
```{r, results = 'asis'}
# child method variable
assign("method", "chiSqGxE", child_env)
snps <- c("11:47328213:T:C", "3:100037644:A:G", "2:10961902:T:C", "11:120531547:C:G","11:46574459:T:A", "2:217410559:G:A", "6:138766021:C:T", "9:136601110:C:G", "15:33122966:C:T")
# snps <- c("11:47328213:T:C")
snps_out <- sort(paste0("chr", gsub(":", "_", snps)))

# Call child Rmd script
res <- lapply(snps_out, function(snp) {
  assign("snp", snp, child_env)
  knitr::knit_child('posthoc/posthoc_child_suggestive_gxe_bmi.Rmd', envir = child_env, quiet = TRUE)
  }
)

cat(unlist(res), sep = '\n')
```


# 2DF/3DF findings

## Notes
- with the exception of 15:33122966:C:T, these SNPs are associated with CRC and/or BMI  
- using 2DF finding for chromosome 5 marker (5:134486618:G:A vs 5:134488349:G:T, the latter is not found in 1000G)  

```{r, results = 'asis'}
# child method variable
assign("method", "chiSqGxE", child_env)
snps <- c("20:6584196:C:G", "15:33122966:C:T", "15:32987718:A:G", "5:134486618:G:A", "10:114274269:T:C", "12:50177324:A:T")
snps_out <- sort(paste0("chr", gsub(":", "_", snps)))

# Call child Rmd script
res <- lapply(snps_out, function(snp) {
  assign("snp", snp, child_env)
  knitr::knit_child('posthoc/posthoc_child_suggestive_gxe_bmi.Rmd', envir = child_env, quiet = TRUE)
  }
)

cat(unlist(res), sep = '\n')
```







# Two-step (expectation based partitioning)
```{r, results = 'asis'}
# child method variable
assign("method", "chiSqGxE", child_env)
snps_out <- c("chr15_33122966_C_T", "chr15_33120215_T_C")

# Call child Rmd script
res <- lapply(snps_out, function(snp) {
  assign("snp", snp, child_env)
  knitr::knit_child('posthoc/posthoc_child_bmi.Rmd', envir = child_env, quiet = TRUE)
  }
)

cat(unlist(res), sep = '\n')
```



