---
title: "`r params$exposure` posthoc analysis"
date: '`r paste("Updated on", Sys.Date())`'
output: 
  html_document:
    css: posthoc/posthoc_report.css
    toc: true
    toc_float:
      collapsed: true
      smooth_scroll: false
    toc_depth: 3
    theme: "paper"
params:
  exposure: "bmi5"
  covariates: !r c("age_ref_imp", "sex", "pc1", "pc2", "pc3", "study_gxe")
  hrc_version: "v2.3"
  path: "/media/work/gwis_test"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, message = F, warning = F, error = F)

# load packages and define wrapper functions
library(tidyverse)
library(data.table)
library(forcats)
library(ggplot2)
library(qqman)
library(rlang)
library(figifs)
library(DT)
library(grid)
library(stargazer)
library(kableExtra)
library(formattable)
library(glue)
library(qs)


exposure = "bmi5"
covariates = c("age_ref_imp", "sex", "pc1", "pc2", "pc3", "study_gxe")
hrc_version = "v2.3"
path = glue("/media/work/gwis_test/{exposure}/")
```

# Description

## Follow-up analyses of GWIS findings  
- Table of significant and suggestive (1df GxE only) ```r params$exposure``` GWIS findings  
- GxE models stratified by study_design and tumor site (proximal, distal, rectal)  
- GxE models with additional covariate sets (might do this for aspirin ..)  
- Odds ratios stratified by exposure and genotype  
- Plots of model predicted outcome by 1 unit increase of exposure/allelic dosage  
- Locuszoom plot   
- Functional annotation plot   

---

<!-- read script containing useful chunks -->
```{r child = 'posthoc/chunks.Rmd'}
```

# Suggestive GxE findings
```{r, results = 'asis'}
statistic = 'chiSqGxE'

# list of SNPs
snps <- sort(c("11:47328213:T:C", "3:100037644:A:G", "2:10961902:T:C", "11:120531547:C:G","11:46574459:T:A", "2:217410559:G:A", "6:138766021:C:T", "9:136601110:C:G", "15:33122966:C:T"))
snps <- paste0('chr', gsub(":", "_", snps))

# see chunks.Rmd for names
run_functions <- function(snp) {
  <<title>>
  <<allele_freq_plots>>
  <<gxe_stratified_sex>>
  <<gxe_stratified_study>>
  <<gxe_stratified_tumorsite>>
  <<stratified_or_cat_bmic3f>>
  <<stratified_or_cat_bmic3f_smk_ever>>
  <<locuszoom_plots>>
  <<functional_plots>>
}

# run
walk(snps, ~ run_functions(.x))
```


# Two-step findings
```{r, results = 'asis'}
statistic = 'chiSqGxE'

# list of SNPs
snps <- sort(c("15:33122966:C:T", "15:33120215:T:C"))
snps <- paste0('chr', gsub(":", "_", snps))

# see chunks.Rmd for names
run_functions <- function(snp) {
  <<title>>
  <<allele_freq_plots>>
  <<gxe_stratified_sex>>
  <<gxe_stratified_study>>
  <<gxe_stratified_tumorsite>>
  <<stratified_or_cat_bmic3f>>
  <<stratified_or_cat_bmic3f_smk_ever>>
  <<locuszoom_plots>>
  <<functional_plots>>
}

# run
walk(snps, ~ run_functions(.x))
```




# 3DF findings

All SNPs associated with BMI, no evidence of GxE (pvalue > 0.05), very small evidence of main effects


```{r, results = 'asis'}
statistic = 'chiSq3df'

# list of SNPs
snps <- sort(c("1:177889480:A:G", "12:50204089:G:A", "16:28649651:C:A", "16:53800954:T:C","18:57850583:C:A", "18:57971625:A:G", "18:58083923:A:C", "19:47581242:C:T","2:417167:T:C", "2:558080:G:T", "2:651365:G:A", "4:45181334:A:T"))
snps <- paste0('chr', gsub(":", "_", snps))

# see chunks.Rmd for names
run_functions <- function(snp) {
  <<title>>
  # <<allele_freq_plots>>
  # <<gxe_stratified_sex>>
  # <<gxe_stratified_study>>
  # <<gxe_stratified_tumorsite>>
  # <<stratified_or_cat_bmic3f>>
  # <<stratified_or_cat_bmic3f_smk_ever>>
  <<locuszoom_plots>>
  <<functional_plots>>
}

# run
walk(snps, ~ run_functions(.x))
```





# 3DF finding (rs7313400)

All SNPs associated with BMI, no evidence of GxE (pvalue > 0.05), very small evidence of main effects


```{r, results = 'asis'}
statistic = 'chiSq3df'

# list of SNPs
snps <- sort(c("12:50204089:G:A"))
snps <- paste0('chr', gsub(":", "_", snps))

# see chunks.Rmd for names
run_functions <- function(snp) {
  <<title>>
  # <<allele_freq_plots>>
  # <<gxe_stratified_sex>>
  # <<gxe_stratified_study>>
  # <<gxe_stratified_tumorsite>>
  <<interaction_plots_bmic3f>>
  <<stratified_or_cat_bmic3f>>
  <<stratified_or_cat_bmic3f_smk_ever>>
  <<locuszoom_plots>>
  <<functional_plots>>
}

# run
walk(snps, ~ run_functions(.x))
```



