---
title: "`r params$exposure` posthoc analysis"
date: '`r paste("Updated on", Sys.Date())`'
output: 
  html_document:
    css: posthoc/posthoc_report.css
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
    toc_depth: 3
    theme: "paper"
params:
  exposure: "methrswklns"
  covariates: !r c("age_ref_imp", "sex", "pc1", "pc2", "pc3", "energytot_imp", "study_gxe")
  hrc_version: "v3.1"
  path: "/media/work/gwis_test"
---

  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, message = F, warning = F, error = F)
# load packages and define wrapper functions
library(tidyverse)
library(data.table)
library(forcats)
library(ggplot2)
library(qqman)
library(rlang)
library(figifs)
library(DT)
library(grid)
library(stargazer)
library(kableExtra)
library(formattable)
library(glue)


# parameters
exposure = params$exposure
covariates = c("age_ref_imp",
               "sex",
               "energytot_imp",
               "pc1",
               "pc2",
               "pc3",
               "study_gxe")
hrc_version = params$hrc_version
path = glue("/media/work/gwis_test/{params$exposure}")
```

# Description

## Follow-up analyses of GWIS findings  
- Table of significant and suggestive (1df GxE only) ```r params$exposure``` GWIS findings  
- GxE models stratified by study_design and tumor site (proximal, distal, rectal)  
- GxE models with additional covariate sets (might do this for aspirin ..)  
- Odds ratios stratified by exposure and genotype  
- Plots of model predicted outcome by 1 unit increase of exposure/allelic dosage  
- Locuszoom plot   
- Functional annotation plot   

---

<!-- read script containing useful chunks -->
```{r child = 'posthoc/chunks.Rmd'}
```

# Case-only finding

Stratified tables show methrswklns / 10 variable  

Stratified OR table uses methrswklns / 10, scaled by the median study/sex specific IQR. Reference group is overall median value, comparison group is +1 overall IQR  

```{r, results = 'asis'}
statistic = 'chiSqCase'

# list of SNPs
snps <- c("21:44227260:T:C")
snps <- paste0('chr', gsub(":", "_", snps))

# see chunks.Rmd for names
run_functions <- function(snp) {
  <<title>>
  <<allele_freq_plots>>
  <<gxe_stratified_sex_v2>>
  <<gxe_stratified_study_v2>>
  <<gxe_stratified_tumorsite_v2>>
  <<gxe_stratified_bmic3_v2>>
  <<gxe_stratified_sex_v2_crc>>
  <<gxe_stratified_study_v2_crc>>
  <<gxe_stratified_tumorsite_v2_crc>>
  <<gxe_stratified_bmic3_v2_crc>>
  <<stratified_or_cont>>
  <<locuszoom_plots>>
  <<functional_plots>>
}

# run
walk(snps, ~ run_functions(.x))
```





<!-- # Case-only finding -->
<!-- ```{r, results = 'asis', eval = exposure == "methrswklns"} -->
<!-- cat("methrswklns was divided by 10 and scaled by the median sex/study specific IQR. For analysis, the reference group is the overall median methrswklns level. That is, reference group (0) is the overall median methrswklns, comparison group (1) is median + 1 overall methrswklns IQR)") -->
<!-- ``` -->

<!-- ```{r, results = 'asis'} -->
<!-- assign("method", "chiSqCase", child_env) -->

<!-- snps_out <- paste0('chr', gsub(":", "_", c("21:44227260:T:C"))) -->

<!-- res <- lapply(snps_out, function(snp) { -->
<!--   assign("snp", snp, child_env) -->
<!--   knitr::knit_child('posthoc/posthoc_child.Rmd', envir = child_env, quiet = TRUE) -->
<!--   } -->
<!-- ) -->

<!-- cat(unlist(res), sep = '\n') -->
<!-- ``` -->
