
```{r include = F}

twostep_wrapper <- function(exposure, step1_statistic, filename_suffix = "") {
  knitr::include_graphics(glue("/media/work/gwis/posthoc/", exposure, "/twostep_wht_", step1_statistic, "_", exposure, filename_suffix, ".png"))
}


# some modifications compared to those of 1df tests
significant_results_table_wrapper <- function(exposure, step1_statistic, filename_suffix = "") {
  
  x <- readRDS(glue("/media/work/gwis/posthoc/", exposure, "/twostep_wht_", step1_statistic, "_", exposure, filename_suffix, "_df.rds")) %>% 
    dplyr::rename(Bin = bin_number) %>% 
    dplyr::mutate(Step1p = formatC(step1p, format = "e", digits = 4),
           Step2p = formatC(step2p, format = "e", digits = 4)) %>% 
    dplyr::arrange(Chromosome, Location) %>%
    dplyr::select(SNP, Subjects, Cases, Bin, Step1p, Step2p)
  
  datatable(x, filter = 'top', options = list(pageLength = 10, autoWidth = T))
}

```


### Step1 D|G
```{r , echo=FALSE, out.width = '100%'}
twostep_wrapper(exposure = params$exposure, step1_statistic = 'chiSqG')
significant_results_table_wrapper(exposure = params$exposure, step1_statistic = 'chiSqG')
```

### Step1 D|G (no GWAS)
```{r , echo=FALSE, out.width = '100%'}
twostep_wrapper(exposure = params$exposure, step1_statistic = 'chiSqG', filename_suffix = '_no_gwas')
significant_results_table_wrapper(exposure = params$exposure, step1_statistic = 'chiSqG', filename_suffix = '_no_gwas')
```

### Step1 G|E
```{r , echo=FALSE, out.width = '100%'}
twostep_wrapper(exposure = params$exposure, step1_statistic = 'chiSqGE')
significant_results_table_wrapper(exposure = params$exposure, step1_statistic = 'chiSqGE')
```

### Step1 G|E (no GWAS)
```{r , echo=FALSE, out.width = '100%'}
twostep_wrapper(exposure = params$exposure, step1_statistic = 'chiSqGE', filename_suffix = '_no_gwas')
significant_results_table_wrapper(exposure = params$exposure, step1_statistic = 'chiSqGE', filename_suffix = '_no_gwas')
```

### Step1 EDGE
```{r , echo=FALSE, out.width = '100%'}
twostep_wrapper(exposure = params$exposure, step1_statistic = 'chiSqEDGE')
significant_results_table_wrapper(exposure = params$exposure, step1_statistic = 'chiSqEDGE')
```

### Step1 EDGE (no GWAS)
```{r , echo=FALSE, out.width = '100%'}
twostep_wrapper(exposure = params$exposure, step1_statistic = 'chiSqEDGE', filename_suffix = '_no_gwas')
significant_results_table_wrapper(exposure = params$exposure, step1_statistic = 'chiSqEDGE', filename_suffix = '_no_gwas')
```

