---
title: "`r params$exposure` posthoc analysis"
date: '`r paste("Updated on", Sys.Date())`'
output: 
  html_document:
    css: ./posthoc/posthoc_report.css
    toc: true
    toc_float:
      collapsed: true
      smooth_scroll: false
    toc_depth: 3
    theme: "paper"
params:
  exposure: "vegetable5qcm"
  covariates: !r c("age_ref_imp", "sex", "energytot_imp", "pc1", "pc2", "pc3", "study_gxe")
  hrc_version: "v3.0"
  path: "/media/work/gwis_test"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, message = F, warning = F, error = F)

# load packages and define wrapper functions
library(tidyverse)
library(data.table)
library(forcats)
library(ggplot2)
library(qqman)
library(rlang)
library(figifs)
library(DT)
library(grid)
library(stargazer)
library(kableExtra)
library(formattable)
library(glue)
library(qs)
```


# Description

## Follow-up analyses of GWIS findings  
- Table of significant and suggestive (1df GxE only) ```r params$exposure``` GWIS findings  
- GxE models stratified by study_design and tumor site (proximal, distal, rectal)  
- GxE models with additional covariate sets (might do this for aspirin ..)  
- Odds ratios stratified by exposure and genotype  
- Plots of model predicted outcome by 1 unit increase of exposure/allelic dosage  
- Locuszoom plot   
- Functional annotation plot   

---

<!-- define variables -->
```{r, include = F}
child_env <- new.env()
assign("exposure",        params$exposure, child_env)
assign("hrc_version",     params$hrc_version, child_env)
assign("covariates", sort(params$covariates), child_env)
# assign("method", "chiSqGxE", child_env)

# exposure working directory path
path = glue("/media/work/gwis_test/{params$exposure}")
```


# 3DF
Output models stratified by sex or tumor site to check for differences in associations
```{r, results = 'asis'}
# child method variable
assign("method", "chiSq3df", child_env)

# SNPs
# snps_clumped <- fread(glue("{path}/data/FIGI_{hrc_version}_gxeset_{exposure}_chiSqGxE_ldclump.clumped"), header = T, stringsAsFactors = F)
# snps_out <- readRDS(glue("{path}/output/manhattan_{exposure}_{hrc_version}_chiSqGxE_df.rds")) %>%
#   dplyr::filter(SNP %in% snps_clumped$SNP,
#                 chiSqGxE_p < 5e-6 | chiSqGxE_p > 5e-8) %>%
#   dplyr::arrange(Chromosome, Location) %>%
#   dplyr::mutate(snps = paste0('chr', gsub("\\:", "\\_", SNP))) %>%
#   dplyr::pull(snps)

snps_out <- "	chr4_91049759_C_G"

# Call child Rmd script
res <- lapply(snps_out, function(snp) {
  assign("snp", snp, child_env)
  knitr::knit_child('./posthoc/posthoc_child.Rmd', envir = child_env, quiet = TRUE)
  })

cat(unlist(res), sep = '\n')
```









<!-- # GxE suggestive -->
<!-- Output models stratified by sex or tumor site to check for differences in associations -->
<!-- ```{r, results = 'asis'} -->
<!-- # child method variable -->
<!-- assign("method", "chiSqGxE", child_env) -->

<!-- # SNPs -->
<!-- snps_clumped <- fread(glue("{path}/data/FIGI_{hrc_version}_gxeset_{exposure}_chiSqGxE_ldclump.clumped"), header = T, stringsAsFactors = F) -->
<!-- snps_out <- readRDS(glue("{path}/output/manhattan_{exposure}_{hrc_version}_chiSqGxE_df.rds")) %>% -->
<!--   dplyr::filter(SNP %in% snps_clumped$SNP, -->
<!--                 chiSqGxE_p < 5e-6 | chiSqGxE_p > 5e-8) %>% -->
<!--   dplyr::arrange(Chromosome, Location) %>% -->
<!--   dplyr::mutate(snps = paste0('chr', gsub("\\:", "\\_", SNP))) %>% -->
<!--   dplyr::pull(snps) -->

<!-- # Call child Rmd script -->
<!-- res <- lapply(snps_out, function(snp) { -->
<!--   assign("snp", snp, child_env) -->
<!--   knitr::knit_child('./posthoc/posthoc_child_suggestive_gxe.Rmd', envir = child_env, quiet = TRUE) -->
<!--   } -->
<!-- ) -->

<!-- cat(unlist(res), sep = '\n') -->
<!-- ``` -->


<!-- # GxE locuszoom plots  -->
<!-- These hits were previously identified as a finding but turn out to be in known GWAS regions and should not be reported as a novel GxE finding -->

<!-- ```{r, results = 'asis'} -->
<!-- # child method variable -->
<!-- assign("method", "chiSqGxE", child_env) -->

<!-- # SNPs -->
<!-- snps_out <- c("chr8_128397494_G_C", "chr6_117816093_A_G") -->


<!-- # Call child Rmd script -->
<!-- res <- lapply(snps_out, function(snp) { -->
<!--   assign("snp", snp, child_env) -->
<!--   knitr::knit_child('./posthoc/posthoc_child_locuszoom_only.Rmd', envir = child_env, quiet = TRUE) -->
<!--   } -->
<!-- ) -->

<!-- cat(unlist(res), sep = '\n') -->
<!-- ``` -->

<!-- # 2DF locuszoom plots -->
<!-- Filtering GWAS + LD SNPs out of the results plots isn't perfect because of slight differences in LD. Ensure that these remaining peaks coincide with known GWAS Loci -->

<!-- ```{r, results = 'asis'} -->
<!-- # child method variable -->
<!-- assign("method", "chiSq2df", child_env) -->


<!-- # SNPs -->
<!-- snps_clumped <- fread(glue("{path}/data/FIGI_{hrc_version}_gxeset_{exposure}_chiSq2df_no_gwas_ldclump.clumped"), header = T, stringsAsFactors = F) -->
<!-- snps_out <- readRDS(glue("{path}/output/manhattan_{exposure}_{hrc_version}_chiSq2df_no_gwas_df.rds")) %>% -->
<!--   dplyr::filter(SNP %in% snps_clumped$SNP) %>% -->
<!--   dplyr::arrange(Chromosome, Location) %>% -->
<!--   dplyr::mutate(snps = paste0('chr', gsub("\\:", "\\_", SNP))) %>% -->
<!--   dplyr::pull(snps) -->

<!-- # Call child Rmd script -->
<!-- res <- lapply(snps_out, function(snp) { -->
<!--   assign("snp", snp, child_env) -->
<!--   knitr::knit_child('./posthoc/posthoc_child_locuszoom_only.Rmd', envir = child_env, quiet = TRUE) -->
<!--   } -->
<!-- ) -->

<!-- cat(unlist(res), sep = '\n') -->
<!-- ``` -->
