
```{r include = F}


# twostep plot
twostep_wrapper <- function(exposure, step1_statistic, eh_method, hrc_version, filename_suffix = "", exclude_gwas = F) {
  
  if(exclude_gwas) {
    filename = glue("{path}/output/twostep_{eh_method}_excludeGWAS_{exposure}_{hrc_version}_{step1_statistic}.png")
  } else {
    filename = glue("{path}/output/twostep_{eh_method}_{exposure}_{hrc_version}_{step1_statistic}.png")
  }
  
  knitr::include_graphics(filename)
}

# table of significant results
significant_results_table_wrapper <- function(exposure, step1_statistic, eh_method, hrc_version, filename_suffix = "", exclude_gwas = F) {
  
  if(exclude_gwas) {
    filename = glue("{path}/output/twostep_{eh_method}_excludeGWAS_{exposure}_{hrc_version}_{step1_statistic}_df.rds")
  } else {
    filename = glue("{path}/output/twostep_{eh_method}_{exposure}_{hrc_version}_{step1_statistic}_df.rds")
  }
  
  # output
  out <- readRDS(filename) %>% 
    dplyr::rename(Bin = bin_number) %>% 
    dplyr::mutate(Step1p = formatC(step1p, format = "e", digits = 4),
           Step2p = formatC(step2p, format = "e", digits = 4)) %>% 
    dplyr::arrange(Chromosome, Location) %>%
    dplyr::select(SNP, Subjects, Cases, Bin, Step1p, Step2p)
  
  out %>% 
    DT::datatable(filter = 'top', options = list(pageLength = 10, autoWidth = T)) %>% 
    DT::formatStyle(columns = colnames(out), fontSize = '75%')
}



# # twostep plot
# twostep_wrapper <- function(exposure, step1_statistic, eh_method, hrc_version, filename_suffix = "") {
#   knitr::include_graphics(glue("/media/work/gwis_test/{exposure}/output/twostep_{eh_method}_{exposure}_{hrc_version}_{step1_statistic}.png"))
# }
# 
# # table of significant results
# significant_results_table_wrapper <- function(exposure, step1_statistic, eh_method, hrc_version, filename_suffix = "") {
#   
#   # output
#   out <- readRDS(glue("/media/work/gwis_test/{exposure}/output/twostep_{eh_method}_{exposure}_{hrc_version}_{step1_statistic}_df.rds")) %>% 
#     dplyr::rename(Bin = bin_number) %>% 
#     dplyr::mutate(Step1p = formatC(step1p, format = "e", digits = 4),
#            Step2p = formatC(step2p, format = "e", digits = 4)) %>% 
#     dplyr::arrange(Chromosome, Location) %>%
#     dplyr::select(SNP, Subjects, Cases, Bin, Step1p, Step2p)
#   
#   out %>% 
#     DT::datatable(filter = 'top', options = list(pageLength = 10, autoWidth = T)) %>% 
#     DT::formatStyle(columns = colnames(out), fontSize = '75%')
# }

```

### Step1 D|G (Galwey)
```{r , echo=FALSE, out.width = '100%'}
twostep_wrapper(exposure = params$exposure, hrc_version = params$hrc_version, eh_method = "galwey", step1_statistic = 'chiSqG')
significant_results_table_wrapper(exposure = params$exposure, eh_method = "galwey", hrc_version = params$hrc_version, step1_statistic = 'chiSqG')
```

### Step1 G|E (Galwey)
```{r , echo=FALSE, out.width = '100%'}
twostep_wrapper(exposure = params$exposure, hrc_version = params$hrc_version, eh_method = "galwey", step1_statistic = 'chiSqGE')
significant_results_table_wrapper(exposure = params$exposure, eh_method = "galwey", hrc_version = params$hrc_version, step1_statistic = 'chiSqGE')
```

### Step1 EDGE (Galwey)
```{r , echo=FALSE, out.width = '100%'}
twostep_wrapper(exposure = params$exposure, hrc_version = params$hrc_version, eh_method = "galwey", step1_statistic = 'chiSqEDGE')
significant_results_table_wrapper(exposure = params$exposure, eh_method = "galwey", hrc_version = params$hrc_version, step1_statistic = 'chiSqEDGE')
```

### Step1 D|G (Galwey) - Exclude GWAS
```{r , echo=FALSE, out.width = '100%'}
twostep_wrapper(exposure = params$exposure, hrc_version = params$hrc_version, eh_method = "galwey", step1_statistic = 'chiSqG', exclude_gwas = T)
significant_results_table_wrapper(exposure = params$exposure, eh_method = "galwey", hrc_version = params$hrc_version, step1_statistic = 'chiSqG', exclude_gwas = T)
```

### Step1 G|E (Galwey) - Exclude GWAS
```{r , echo=FALSE, out.width = '100%'}
twostep_wrapper(exposure = params$exposure, hrc_version = params$hrc_version, eh_method = "galwey", step1_statistic = 'chiSqGE', exclude_gwas = T)
significant_results_table_wrapper(exposure = params$exposure, eh_method = "galwey", hrc_version = params$hrc_version, step1_statistic = 'chiSqGE', exclude_gwas = T)
```

### Step1 EDGE (Galwey) - Exclude GWAS
```{r , echo=FALSE, out.width = '100%'}
twostep_wrapper(exposure = params$exposure, hrc_version = params$hrc_version, eh_method = "galwey", step1_statistic = 'chiSqEDGE', exclude_gwas = T)
significant_results_table_wrapper(exposure = params$exposure, eh_method = "galwey", hrc_version = params$hrc_version, step1_statistic = 'chiSqEDGE', exclude_gwas = T)
```

