
```{r include = F}

# twostep plot
twostep_wrapper <- function(exposure, step1_statistic, hrc_version, filename_suffix = "") {
  knitr::include_graphics(glue("/media/work/gwis_test/{exposure}/output/twostep_{exposure}_{hrc_version}_{step1_statistic}.png"))
}

# table of significant results
significant_results_table_wrapper <- function(exposure, step1_statistic, hrc_version, filename_suffix = "") {
  
  # output
  out <- readRDS(glue("/media/work/gwis_test/{exposure}/output/twostep_{exposure}_{hrc_version}_{step1_statistic}_df.rds")) %>% 
    dplyr::rename(Bin = bin_number) %>% 
    dplyr::mutate(Step1p = formatC(step1p, format = "e", digits = 4),
           Step2p = formatC(step2p, format = "e", digits = 4)) %>% 
    dplyr::arrange(Chromosome, Location) %>%
    dplyr::select(SNP, Subjects, Cases, Bin, Step1p, Step2p)
  
  out %>% 
    DT::datatable(filter = 'top', options = list(pageLength = 10, autoWidth = T)) %>% 
    DT::formatStyle(columns = colnames(out), fontSize = '75%')
}

```

### Step1 D|G
```{r , echo=FALSE, out.width = '100%'}
twostep_wrapper(exposure = params$exposure, hrc_version = params$hrc_version, step1_statistic = 'chiSqG')
significant_results_table_wrapper(exposure = params$exposure, hrc_version = params$hrc_version, step1_statistic = 'chiSqG')
```

<!-- ### Step1 D|G (no GWAS) -->
<!-- ```{r , echo=FALSE, out.width = '100%'} -->
<!-- twostep_wrapper(exposure = params$exposure, step1_statistic = 'chiSqG', filename_suffix = '_no_gwas') -->
<!-- significant_results_table_wrapper(exposure = params$exposure, step1_statistic = 'chiSqG', filename_suffix = '_no_gwas') -->
<!-- ``` -->

### Step1 G|E
```{r , echo=FALSE, out.width = '100%'}
twostep_wrapper(exposure = params$exposure, hrc_version = params$hrc_version, step1_statistic = 'chiSqGE')
significant_results_table_wrapper(exposure = params$exposure, hrc_version = params$hrc_version, step1_statistic = 'chiSqGE')
```

<!-- ### Step1 G|E (no GWAS) -->
<!-- ```{r , echo=FALSE, out.width = '100%'} -->
<!-- twostep_wrapper(exposure = params$exposure, step1_statistic = 'chiSqGE', filename_suffix = '_no_gwas') -->
<!-- significant_results_table_wrapper(exposure = params$exposure, step1_statistic = 'chiSqGE', filename_suffix = '_no_gwas') -->
<!-- ``` -->

### Step1 EDGE
```{r , echo=FALSE, out.width = '100%'}
twostep_wrapper(exposure = params$exposure, hrc_version = params$hrc_version, step1_statistic = 'chiSqEDGE')
significant_results_table_wrapper(exposure = params$exposure, hrc_version = params$hrc_version, step1_statistic = 'chiSqEDGE')
```

<!-- ### Step1 EDGE (no GWAS) -->
<!-- ```{r , echo=FALSE, out.width = '100%'} -->
<!-- twostep_wrapper(exposure = params$exposure, step1_statistic = 'chiSqEDGE', filename_suffix = '_no_gwas') -->
<!-- significant_results_table_wrapper(exposure = params$exposure, step1_statistic = 'chiSqEDGE', filename_suffix = '_no_gwas') -->
<!-- ``` -->

