---
title: "FIGI CORECT Preimputation Quality Control"
bibliography: /home/rak/Dropbox/library.bib
output: 
  #html_document
  word_document:
    reference_docx: /home/rak/Dropbox/FIGI/Code/Functions/word_styles.docx
---

<style type="text/css">
.main-container {
  max-width: 1800px !important;
  margin-left: auto;
  margin-right: auto;
}
</style>

# Goals

Document GWAS/Pre-imputation quality control steps for genotype imputation on Haplotype Reference Consortium reference panel via the Michigan Imputation Server.  

Most of the GWAS QC was performed by Chris Edlund. These files then served as starting points for pre-imputation quality control, which mainly involves ensuring that Ref/Alt alleles are properly assigned, removing monomorphic and ambiguous SNPs (look up what else it does) -- steps are outlined in the Michigan Imputation Server [website](https://imputationserver.sph.umich.edu/index.html#!pages/help), and are optional but recommended to help identify issues with GWAS input data.  

I used either of the scripts provided in the Michigan Imputation Server to perform pre-imputation QC:   
* [checkVCF](https://github.com/zhanxw/checkVCF)  
* [HRC-1000G-check-bim.pl](https://www.well.ox.ac.uk/~wrayner/tools/)  

------

# GWAS QC

You can find documentation generated by Chris Edlund in the same folder as this document 

------

# Pre-Imputation QC

## axiom_acs_aus_nf

Starting files were pre-phased 06/13/2013 in Oxford phased haplotype format i.e. SHAPEIT (haps/sample):  
(Location in cedlund/corect folder in pmd-01)
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr10.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr11.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr12.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr13.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr14.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr15.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr16.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr17.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr18.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr19.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr1.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr20.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr21.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr22.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr2.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr3.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr4.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr5.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr6.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr7.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr8.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr9.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chrX_nonPAR.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chrX_PAR1.haps  
- axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chrX_PAR2.haps   

### Pre-imputation steps:
```{bash, eval=F} 
# Step 01 - insert chromosome information to each file
awk '$1=1' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr1.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr1.haps
awk '$1=2' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr2.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr2.haps
awk '$1=3' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr3.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr3.haps
awk '$1=4' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr4.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr4.haps
awk '$1=5' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr5.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr5.haps
awk '$1=6' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr6.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr6.haps
awk '$1=7' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr7.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr7.haps
awk '$1=8' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr8.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr8.haps
awk '$1=9' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr9.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr9.haps
awk '$1=10' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr10.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr10.haps
awk '$1=11' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr11.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr11.haps
awk '$1=12' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr12.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr12.haps
awk '$1=13' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr13.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr13.haps
awk '$1=14' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr14.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr14.haps
awk '$1=15' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr15.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr15.haps
awk '$1=16' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr16.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr16.haps
awk '$1=17' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr17.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr17.haps
awk '$1=18' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr18.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr18.haps
awk '$1=19' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr19.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr19.haps
awk '$1=20' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr20.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr20.haps
awk '$1=21' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr21.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr21.haps
awk '$1=22' axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr22.haps > axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr22.haps

# Step 02 - create sample files for each chromosome
# (need to create one for each chromosome, was not available)
for i in `seq 1 22`;
do
    cp axiom_corect_acs_aus_nf_pre-impute_2013-06-13_chr22.sample axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr$i.sample
done

# Step03 - convert to VCF format
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr1 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr1.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr2 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr2.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr3 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr3.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr4 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr4.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr5 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr5.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr6 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr6.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr7 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr7.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr8 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr8.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr9 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr9.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr10 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr10.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr11 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr11.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr12 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr12.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr13 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr13.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr14 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr14.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr15 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr15.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr16 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr16.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr17 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr17.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr18 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr18.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr19 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr19.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr20 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr20.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr21 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr21.vcf
shapeit -convert --input-haps axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr22 --output-vcf axiom_corect_acs_aus_nf_phased_hrc_2017-01-31_chr22.vcf

# Step04 - sort and bgzip
for i in {1..22}
do
    vcf-sort axiom_corect_acs_aus_nf_phased_hrc_bcftools_2017-01-31_chr$i.vcf | bgzip -c > axiom_corect_acs_aus_nf_phased_hrc_bcftools_2017-01-31_chr$i.vcf.gz
done

# Step05 - run vcfcheck.pl script
# (upladed these QC files to the Michigan Imputation Server)
for i in {1..22}
do
	~/scripts/checkVCF.py -r ~/scripts/human_g1k_v37.fasta -o chr$i axiom_corect_acs_aus_nf_phased_hrc_bcftools_2017-01-31_chr$i.vcf.gz
done
```
<br><br><br><br>

## axiom_mecc_cfr_ky
Starting files were pre-phased 01/23/2013 in Oxford phased haplotype format i.e. SHAPEIT (haps/sample):  
(Location in cedlund/corect folder in pmd-01)
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr1.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr2.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr3.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr4.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr5.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr6.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr7.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr8.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr9.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr10.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr11.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr12.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr13.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr14.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr15.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr16.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr17.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr18.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr19.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr20.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr21.haps  
- axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr22.haps 

```{bash, eval=F} 

# Step 01 - insert chromosome information to each file
awk '$1=1' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr1.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr1.haps
awk '$1=2' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr2.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr2.haps
awk '$1=3' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr3.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr3.haps
awk '$1=4' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr4.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr4.haps
awk '$1=5' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr5.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr5.haps
awk '$1=6' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr6.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr6.haps
awk '$1=7' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr7.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr7.haps
awk '$1=8' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr8.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr8.haps
awk '$1=9' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr9.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr9.haps
awk '$1=10' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr10.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr10.haps
awk '$1=11' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr11.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr11.haps
awk '$1=12' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr12.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr12.haps
awk '$1=13' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr13.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr13.haps
awk '$1=14' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr14.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr14.haps
awk '$1=15' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr15.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr15.haps
awk '$1=16' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr16.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr16.haps
awk '$1=17' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr17.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr17.haps
awk '$1=18' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr18.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr18.haps
awk '$1=19' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr19.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr19.haps
awk '$1=20' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr20.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr20.haps
awk '$1=21' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr21.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr21.haps
awk '$1=22' axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23_chr22.haps > axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr22.haps

# Step 02 - create sample files for each chromosome
# (need to create one for each chromosome, was not available)
for i in `seq 1 22`;
do
        cp axiom_corect_mecc_cfr_ky_pre-impute_2013-01-23.sample  axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr$i.sample
done

# Step03 - convert to VCF format
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr1 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr1.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr2 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr2.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr3 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr3.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr4 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr4.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr5 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr5.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr6 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr6.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr7 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr7.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr8 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr8.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr9 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr9.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr10 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr10.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr11 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr11.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr12 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr12.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr13 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr13.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr14 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr14.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr15 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr15.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr16 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr16.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr17 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr17.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr18 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr18.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr19 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr19.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr20 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr20.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr21 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr21.vcf
shapeit -convert --input-haps axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr22 --output-vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr22.vcf

# Step04 - fix reference allele based on 1000G fasta file 
# (in retrospect, this step was not necessary since the purpose of the script was simply to check VCF files for validity, and since Michigan Imputation Server conducts QC anyway. Including as part of documentation)
# (see "help" page in the michigan imputation server website)
for i in {1..22}
do
  bcftools norm -c s -f ~/scripts/human_g1k_v37.fasta -o axiom_corect_mecc_cfr_ky_phased_hrc_bcftools_2017-02-03_chr$i.vcf axiom_corect_mecc_cfr_ky_phased_hrc_2017-02-03_chr$i.vcf
done

# Step05 - sort and bgzip
for i in {1..22}
do
        vcf-sort axiom_corect_mecc_cfr_ky_phased_hrc_bcftools_2017-02-03_chr$i.vcf | bgzip -c > axiom_corect_mecc_cfr_ky_phased_hrc_bcftools_2017-02-03_chr$i.vcf.gz
done

# Step06 - run vcfcheck.pl script again
# (upladed these QC files to the Michigan Imputation Server)
for i in {1..22}
do
  ~/scripts/checkVCF.py -r ~/scripts/human_g1k_v37.fasta -o chr$i axiom_corect_mecc_cfr_ky_phased_hrc_bcftools_2017-02-03_chr$i.vcf.gz
done
```
<br><br><br><br>



## ccfr_1m_1mduo

Input file (binary plink format):  
* ccfr_1M-1Mduo_impute_20120614  

This batch was reimputed after dropping samples with low call rates to avoid missing chunks in output file. In previous attempt, received this error message:  
```{bash, eval=F}
chunk_14_0000000001_0020000000 Sample 160000032691_160000032691: call rate: 0.3333333333333333
chunk_14_0000000001_0020000000 Sample 160000038102_160000038102: call rate: 0.3333333333333333
chunk_14_0000000001_0020000000 (Snps: 3, Reference overlap: 1.0, low sample call rates: true)
```

This chunk from chr14 was omitted. To fix, remove samples 160000032691_160000032691 + 160000038102_160000038102. 

Create exclusion list for plink:  
```{r, eval=F}
# exclude the following:
#160000032691_160000032691
#160000038102_160000038102

# just output a text file for it
setwd("~/git/ccfr_1m_1mduo_reimpute/")

df <- data.frame(FID = c("160000032691", "160000038102"), IID = c("160000032691", "160000038102"))

write.table(df, file = "~/ccfr_1m_drop_samples.txt", quote = F, row.names = F, col.names = F, sep = '\t')
system("scp ~/ccfr_1m_drop_samples.txt hpcc:/staging/dvc/andreeki/ccfr_1m")
system("rm ~/ccfr_1m_drop_samples.txt")
```

Remove samples, then QC using **HRC-1000G-check-bim.pl** script:  
```{bash, eval=F}
WDIR=/staging/dvc/andreeki
ODIR=/auto/pmd-02/figi/Preimputation_QC_Plink/ccfr_1m_1mduo
REF=/auto/pmd-02/figi/andreeki/Refs

mkdir -p ${WDIR}/ccfr_1m
cd ${WDIR}/ccfr_1m

run_wrayner () {
echo "#!/bin/bash
#SBATCH --time=200:00:00
#SBATCH --ntasks=1
#SBATCH --mem=64GB
#SBATCH --account=lc_dvc
#SBATCH --partition=conti
#SBATCH --job-name=ccfr_1m_run_wrayner
#SBATCH --output=ccfr_1m_run_wrayner.log

plink --bfile ${ODIR}/ccfr_1M-1Mduo_impute_20120614 \
      --remove ccfr_1m_drop_samples.txt \
      --make-bed \
      --out ccfr_1m_1mduo_update

plink --bfile $1 --freq --nonfounders --out $1

perl ${REF}/HRC-1000G-check-bim-NoReadKey.pl \
    -b $1.bim \
    -f $1.frq \
    -r ${REF}/HRC.r1-1.GRCh37.wgs.mac5.sites.tab -h -n" | sbatch

}

run_wrayner "ccfr_1m_1mduo_update"
```

Then convert files to VCF, sort, bgzip, and upload:  
```{bash, eval=F}
#!/bin/bash
#SBATCH --time=200:00:00
#SBATCH --ntasks=1
#SBATCH --mem=32GB
#SBATCH --account=lc_dvc
#SBATCH --partition=conti
#SBATCH --job-name=vm.wrayner2
#SBATCH --output=vm.wrayner2.log

cd $SLURM_SUBMIT_DIR

# Here - should have used vcf-iid , but keep as is for consistency with previous effort..
# FYI - the reason this didn't happen with other batches is that I converted shapeit files into vcf.gz. 
for chr in {1..22}
do
    plink --bfile  FIGI_tmp-updated-chr$chr --recode vcf --out FIGI-updated-chr$chr
    vcf-sort FIGI-updated-chr$chr.vcf | bgzip -c > FIGI-updated-chr$chr.vcf.gz
done
```
<br><br><br><br>



## ccfr_omni

Input file (binary plink format):  
* ccfr_Omni_cases-and-controls_impute_20120614  

QC using **HRC-1000G-check-bim.pl** script:  
```{bash, eval = F}
#!/bin/bash
#PBS -l nodes=1:ppn=1
#PBS -l walltime=23:00:00
#PBS -l mem=32gb
#PBS -l arch=x86_64
#PBS -A lc_dvc
#PBS -q conti

cd $PBS_O_WORKDIR

#source /auto/usc/perl/5.20.0/setup.sh
plink --bfile ccfr_Omni_cases-and-controls_impute_20120614 --freq --out ccfr_Omni_cases-and-controls_impute_20120614

perl /staging/dvc/andreeki/refs/HRC-1000G-check-bim.pl -b ccfr_Omni_cases-and-controls_impute_20120614.bim -f ccfr_Omni_cases-and-controls_impute_20120614.frq -r /staging/dvc/andreeki/refs/HRC.r1-1.GRCh37.wgs.mac5.sites.tab -n -h

```

Run plink.sh output from above (output updated binary plink files):
```{bash, eval=F}
#!/bin/bash
#PBS -l nodes=1:ppn=1
#PBS -l walltime=23:00:00
#PBS -l mem=32gb
#PBS -l arch=x86_64
#PBS -A lc_dvc
#PBS -q conti

cd $PBS_O_WORKDIR

chmod +x Run-plink.sh
./Run-plink.sh

```


Convert files to VCF, sort, bgzip, and upload:  
```{bash, eval = F}
#!/bin/bash
#PBS -l nodes=1:ppn=1
#PBS -l walltime=23:00:00
#PBS -l mem=32gb
#PBS -l arch=x86_64
#PBS -A lc_dvc
#PBS -q conti

cd $PBS_O_WORKDIR

for i in {1..22}
do
        plink --bfile  ccfr_Omni_cases-and-controls_impute_20120614-updated-chr$i --recode vcf --a2-allele Force-Allele1-ccfr_Omni_cases-and-controls_impute_20120614-HRC.txt 2 1 --out ccfr_Omni_cases-and-controls_impute_20120614-updated-chr$i

done
```

(also ran the alternate QC just in case)
```{bash, eval = F}
for i in {1..22}
do

echo "
#!/bin/bash
#PBS -l nodes=1:ppn=1
#PBS -l walltime=23:00:00
#PBS -l mem=32gb
#PBS -l arch=x86_64
#PBS -A lc_dvc
#PBS -q conti

cd \$PBS_O_WORKDIR

vcf-sort ccfr_Omni_cases-and-controls_impute_20120614-updated-chr$i.vcf | bgzip -c > ccfr_Omni_cases-and-controls_impute_20120614-updated-chr$i.vcf.gz

/staging/dvc/andreeki/refs/checkVCF.py -r /staging/dvc/andreeki/refs/human_g1k_v37.fasta -o Omni_chr$i ccfr_Omni_cases-and-controls_impute_20120614-updated-chr$i.vcf.gz
" > step04_omni_chr$i.pbs

done


for i in {1..22}
do
qsub step04_omni_chr$i.pbs
done

```
<br><br><br><br>

## corect_oncorrray

Input files are on bioghost server (biolinuxlogin2.usc.edu) @ /FIGI/oncoarray/imputation_2016-03-18/post-imputation/prephased:  
- corect_oncoarray_release_v3_phased_chr10.haps  
- corect_oncoarray_release_v3_phased_chr10.sample  
- corect_oncoarray_release_v3_phased_chr11.haps    
- corect_oncoarray_release_v3_phased_chr11.sample  
- corect_oncoarray_release_v3_phased_chr12.haps  
- corect_oncoarray_release_v3_phased_chr12.sample  
- corect_oncoarray_release_v3_phased_chr13.haps  
- corect_oncoarray_release_v3_phased_chr13.sample  
- corect_oncoarray_release_v3_phased_chr14.haps  
- corect_oncoarray_release_v3_phased_chr14.sample  
- corect_oncoarray_release_v3_phased_chr15.haps  
- corect_oncoarray_release_v3_phased_chr15.sample  
- corect_oncoarray_release_v3_phased_chr16.haps  
- corect_oncoarray_release_v3_phased_chr16.sample  
- corect_oncoarray_release_v3_phased_chr17.haps  
- corect_oncoarray_release_v3_phased_chr17.sample  
- corect_oncoarray_release_v3_phased_chr18.haps  
- corect_oncoarray_release_v3_phased_chr18.sample  
- corect_oncoarray_release_v3_phased_chr19.haps  
- corect_oncoarray_release_v3_phased_chr19.sample  
- corect_oncoarray_release_v3_phased_chr1.haps  
- corect_oncoarray_release_v3_phased_chr1.sample  
- corect_oncoarray_release_v3_phased_chr20.haps  
- corect_oncoarray_release_v3_phased_chr20.sample  
- corect_oncoarray_release_v3_phased_chr21.haps  
- corect_oncoarray_release_v3_phased_chr21.sample  
- corect_oncoarray_release_v3_phased_chr22.haps  
- corect_oncoarray_release_v3_phased_chr22.sample  
- corect_oncoarray_release_v3_phased_chr2.haps  
- corect_oncoarray_release_v3_phased_chr2.sample  
- corect_oncoarray_release_v3_phased_chr3.haps  
- corect_oncoarray_release_v3_phased_chr3.sample  
- corect_oncoarray_release_v3_phased_chr4.haps  
- corect_oncoarray_release_v3_phased_chr4.sample  
- corect_oncoarray_release_v3_phased_chr5.haps  
- corect_oncoarray_release_v3_phased_chr5.sample  
- corect_oncoarray_release_v3_phased_chr6.haps  
- corect_oncoarray_release_v3_phased_chr6.sample  
- corect_oncoarray_release_v3_phased_chr7.haps  
- corect_oncoarray_release_v3_phased_chr7.sample  
- corect_oncoarray_release_v3_phased_chr8.haps  
- corect_oncoarray_release_v3_phased_chr8.sample  
- corect_oncoarray_release_v3_phased_chr9.haps  
- corect_oncoarray_release_v3_phased_chr9.sample  

convert files to VCF
```{bash, eval = F}
for i in {1..22}
do
        ./shapeit -convert --input-haps corect_oncoarray_release_v3_phased_chr$i  --output-vcf corect_oncoarray_release_v3_phased_chr$i.vcf
done
```


run **checkVCF.py** script
(I had trouble running the mrayner perl QC script due to file/memory size constraints)
```{bash, eval = F}
#!/bin/bash

# GOALS:
# 1) fix reference allele basd on 1000g fasta file
# 2) then filter out asian samples
# 3) run vcf-sort, bgzip vcf.gz
# 4) run checkVCF.py script

# this script creates individuals sh files for execution by pbs..

for i in {1..22}
do

echo "
#!/bin/bash

# 1)
bcftools norm -c s -f human_g1k_v37.fasta -o /staging/dvc/andreeki/oncoarray/step02/corect_oncoarray_bcftools_chr$i.vcf \
    /staging/dvc/andreeki/oncoarray/step01/corect_oncoarray_release_v3_phased_chr$i.vcf

# 2)
vcftools --vcf /staging/dvc/andreeki/oncoarray/step02/corect_oncoarray_bcftools_chr$i.vcf --remove asians_to_exclude.txt --recode \
    --out /staging/dvc/andreeki/oncoarray/step02/corect_oncoarray_bcftools_chr$i
mv /staging/dvc/andreeki/oncoarray/step02/corect_oncoarray_bcftools_chr$i.recode.vcf /staging/dvc/andreeki/oncoarray/step02/corect_oncoarray_bcftools_eur_chr$i.vcf


# 3) --- watch folder step03!
vcf-sort /staging/dvc/andreeki/oncoarray/step02/corect_oncoarray_bcftools_eur_chr$i.vcf | \
    bgzip -c > /staging/dvc/andreeki/oncoarray/step03/corect_oncoarray_bcftools_eur_chr$i.vcf.gz

# 4)
./checkVCF.py -r human_g1k_v37.fasta -o /staging/dvc/andreeki/oncoarray/step03/chr$i /staging/dvc/andreeki/oncoarray/step03/corect_oncoarray_bcftools_eur_chr$i.vcf.gz

" > step02_chr$i.sh

chmod u+x step02_chr$i.sh

echo "
#PBS -l walltime=23:00:00
#PBS -l nodes=1:ppn=1
#PBS -l mem=16gb
#PBS -l arch=x86_64
#PBS -A lc_dvc
#PBS -q conti

cd \$PBS_O_WORKDIR
./step02_chr$i.sh
" > step02_chr$i.pbs


done
```

## MECC
Input files:  
- mecc_omni_clean_allmaf_impute_20120718_chr10.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr10.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr11.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr11.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr12.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr12.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr13.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr13.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr14.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr14.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr15.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr15.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr16.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr16.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr17.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr17.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr18.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr18.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr19.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr19.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr1.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr1.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr20.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr20.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr21.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr21.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr22.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr22.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr2.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr2.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr3.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr3.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr4.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr4.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr5.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr5.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr6.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr6.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr7.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr7.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr8.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr8.sample  
- mecc_omni_clean_allmaf_impute_20120718_chr9.haps  
- mecc_omni_clean_allmaf_impute_20120718_chr9.sample  


Convert haps/sample to VCF file format using shapeit
```{bash, eval = F}
#!/bin/bash

# Convert haps/sample to vcf file format using shapeit

#for i in {1..2}
#do
#   echo "$i"
#done



for i in {1..22}
do
    awk '$1 = '"$i"' ' mecc_omni_clean_allmaf_impute_20120718_chr$i.haps > mecc_omni_hrc_20170317_chr$i.haps
    cp mecc_omni_clean_allmaf_impute_20120718_chr$i.sample mecc_omni_hrc_20170317_chr$i.sample
    shapeit -convert --input-haps mecc_omni_hrc_20170317_chr$i  --output-vcf mecc_omni_hrc_20170317_chr$i.vcf
done
```


Use **checkVCF.py** script
```{bash, eval=F }
#!/bin/bash

# GOALS:
# 1) fix reference allele basd on 1000g fasta file
# 2) run vcf-sort, bgzip vcf.gz
# 3) run checkVCF.py script

# this script creates individuals sh files for execution by pbs..

for i in {1..22}
do

echo "
#!/bin/bash

# 1)
bcftools norm -c s -f human_g1k_v37.fasta -o /staging/dvc/andreeki/prephase/mecc_omni_hrc_refalt_20170317_chr$i.vcf \
    /staging/dvc/andreeki/prephase/mecc_omni_hrc_20170317_chr$i.vcf

# 2)
vcf-sort /staging/dvc/andreeki/prephase/mecc_omni_hrc_refalt_20170317_chr$i.vcf | \
    bgzip -c > /staging/dvc/andreeki/prephase/mecc_omni_hrc_refalt_20170317_chr$i.vcf.gz

# 3)
./checkVCF.py -r human_g1k_v37.fasta -o /staging/dvc/andreeki/prephase/chr$i /staging/dvc/andreeki/prephase/mecc_omni_hrc_refalt_20170317_chr$i.vcf.gz

" > step02_chr$i.sh

chmod u+x step02_chr$i.sh

echo "
#PBS -l walltime=23:00:00
#PBS -l nodes=1:ppn=1
#PBS -l mem=32gb
#PBS -l arch=x86_64
#PBS -A lc_dvc
#PBS -q conti

cd \$PBS_O_WORKDIR
./step02_chr$i.sh
" > step02_chr$i.pbs

done
```
